Final CORS fix PR

ljbade opened the pull request:
Squashing into two commits to save time and get it ready for @denniszollo to start testing it ASAP.

ljbade commented:
Need to check has POS flag for race condition
Also push hitl tag

jacobmcnamee commented on libswiftnav:
```diff
@@ -1 +1 @@
-Subproject commit 45657c98c5caba708a7c3116a7e77d4f30f04054
+Subproject commit 428ac469b11f63d9ba71d8c0b1e24be22f527b0e
```
remove libswiftnav update

jacobmcnamee commented on src/solution.c:
```diff
@@ -409,15 +410,34 @@ static void solution_thread(void *arg)
+      /* If we have timing then we can calculate the relationship between
+       * receiver time and GPS time and hence provide the pseudorange
+       * calculation with the local GPS time of reception. */
+      calc_navigation_measurement(n_ready, p_meas, p_nav_meas,
```
Conditionally define the `nav_time` pointer instead of duplicating everything:
`p_nav_time = (time_quality == TIME_FINE) ? &nav_time : NULL;`

jacobmcnamee commented on src/solution.c:
```diff
@@ -440,6 +440,11 @@ static void solution_thread(void *arg)
     ephemeris_unlock();
 
+    for (u32 i=0; i<n_ready; i++) {
+      printf("%d: tot: %.7f rpr: %.7f", nav_meas[i].sid.sat, nav_meas[i].tot.tow,
```
remove `printf()`

ljbade replied:
This was removed in a subsequent commit.

jacobmcnamee commented:
Please remove the libswiftnav pointer updates from intermediate commits. Only include valid pointers to libswiftnav/master.

jacobmcnamee commented on src/base_obs.h:
```diff
@@ -45,6 +45,11 @@ typedef struct {
+/* Maximum distance between calculated and surveyed base station single point
+ * position for error checking.
+ */
+#define BASE_STATION_DISTANCE_THRESHOLD 5
```
units?

jacobmcnamee commented on src/board/v2/error.c:
```diff
@@ -182,7 +182,7 @@ void HardFaultVector(void)
   asm("mrs %0, psp" : "=r"(psp) : : );
   asm("mov %0, lr" : "=r"(lr) : : );
-  
+
```
remove

ljbade replied:
Is there supposed to be a space here? My editor must have automatically removed when I was looking at this file...

jacobmcnamee commented on src/solution.c:
```diff
@@ -403,56 +402,53 @@ static void solution_thread(void *arg)
-                                  (double)((u32)nav_tc)/SAMPLE_FREQ,
-								  &nav_time, p_e_meas);
+                                  (double)((u32)rec_tc) / SAMPLE_FREQ,
+								                  &rec_time, p_e_meas);
```
tabs?

ljbade replied:
Already fixed in subsequent commit.

jacobmcnamee commented on src/solution.c:
```diff
@@ -512,46 +515,58 @@ static void solution_thread(void *arg)
+      nav_meas_tdcp[i].pseudorange += GPS_C * gpsdifftime(&position_solution.time, &rec_time);
     }
-    chMtxUnlock(&base_obs_lock);
+    /**
```
/*

jacobmcnamee commented on src/solution.c:
```diff
@@ -577,8 +577,7 @@ static void solution_thread(void *arg)
       for (u8 i = 0; i < n_ready_tdcp; i++) {
-        nav_meas_tdcp[i].raw_pseudorange -= t_err * nav_meas_tdcp[i].raw_doppler
-                                            * GPS_L1_LAMBDA;
+        nav_meas_tdcp[i].raw_pseudorange -= t_err;
```
squash this change?

ljbade replied:
I made this a separate commit to keep track of who came up with this change.

jacobmcnamee commented on src/track.c:
```diff
@@ -616,6 +624,37 @@ void tracking_channel_measurement_get(tracker_channel_id_t id,
+    tracker_channel->carrier_phase_offset = 0.0;
+  }
+
+  /* Adjust carrier phase initial integer offset to be approximately equal to
```
Remove commented code or combine with commit that uncomments it

jacobmcnamee commented on src/track.c:
```diff
@@ -344,6 +349,9 @@ bool tracker_channel_init(tracker_channel_id_t id, gnss_signal_t sid,
     tracker_channel->elevation = elevation;
 
+    /* Reset the carrier phase offset so it gets set again. */
+    tracker_channel->carrier_phase_offset = 0.0;
```
also set `carrier_phase_offset_lock` ?

jacobmcnamee commented on src/nmea.c:
```diff
@@ -393,22 +393,26 @@ void nmea_gpgll(const gnss_solution *soln, const gps_time_t *gps_t)
-  DO_EVERY(gpvtg_msg_rate,
-    nmea_gpvtg(soln);
-  );
+  if (!skip_velocity) {
```
@denniszollo Is this going to be a problem?

jacobmcnamee replied:
Should we just skip all solution outputs if this happens?

ljbade replied:
When I chatted with @denniszollo last week his preference was to just skip the messages with a velocity..

jacobmcnamee commented on src/solution.c:
```diff
@@ -602,8 +599,15 @@ static void solution_thread(void *arg)
         ephemeris_unlock();
       }
 
-      /* Send the observations. */
```
It looks like this commit is supposed to 
1. still propagate and post back to the mailbox even if `fabs(t_check - (u32)t_check) < TIME_MATCH_THRESHOLD)` fails
2. additionally check for `TIME_FINE` before sending observations

Am I missing something, or is the commit message inaccurate?

ljbade replied:
This confusing due to the way the silly `(u32)t_check` thing is used in firmware from before this PR.

The default settings for Piksi is that `soln_freq` is 10Hz, and `obs_output_divisor` is 2 to give effective rate of 5Hz. Thus you will get a position solution in SBP every 10Hz, and a observation message every 5Hz.

Due to some rearranging of the solution thread code in this PR, the 5Hz check was also covering the position solution code.

So now the time propagation (which is needed for the position solution), and the solution code are run at 5Hz, but the code that sends to observations only runs at 5Hz, and only when the receiver clock has been correctly set (`TIME_FINE`)

jacobmcnamee commented on src/base_obs.c:
```diff
@@ -174,10 +174,7 @@ static void update_obss(obss_t *new_obss)
   /* Copy over the time. */
   base_obss.tor = new_obss->tor;
 
-  /* Reset the `has_pos` flag. */
```
I don't think the change is necessary as all accesses to `base_obss` should be mutex protected.

ljbade replied:
Ah, OK @denniszollo spotted this when we were figuring out the 5Hz stuff.

@denniszollo What is your preference here?

jacobmcnamee commented:
FYI I went through commit by commit, so there are some "comments on outdated diffs"

ljbade commented:
@jacobmcnamee Thanks, ill get to work on them, and the rebase.

jacobmcnamee commented:
Instead of duplicating the lock counter, just clear the carrier phase offset in `tracker_ambiguity_unknown()` (http://github.com/swift-nav/piksi_firmware/blob/master/src/track_api.c#L221)

jacobmcnamee commented:
The sample count for each channel is now managed in the NAP code and simply copied out to the common data struct at each update. See `nap_track_read_results()`. For v2, it would be sufficient to make the sample count (actually called `count_snapshot` now) a u64 as you did in track.c previously. But for v3 the sample count is read out of the NAP each update and not currently stored in software, which is problematic since that value is always a u32 in the NAP (whereas you need the u64 software-extended version).

ljbade commented:
@jacobmcnamee I have rebased on your https://github.com/swift-nav/piksi_firmware/pull/674

I have also used `tracker_ambiguity_unknown()` instead of the lock counter, and moved the offset to the internal struct.

Please review.

jacobmcnamee commented on src/manage.c:
```diff
@@ -678,6 +678,18 @@ static void manage_tracking_startup(void)
       continue;
     }
 
+    /* Transition to tracking. */
```
Remove this. The u64 timing count stuff is unnecessary with the measurement time delta changes.

ljbade replied:
Hmm this might have been an artifact of the rebase

jacobmcnamee commented on src/system_monitor.c:
```diff
@@ -177,11 +173,10 @@ static void system_monitor_thread(void *arg)
 
       if (base_distance > BASE_STATION_DISTANCE_THRESHOLD) {
-        log_warn("Invalid surveyed position coordinates\n");
+        log_warn("Sending invalid surveyed position coordinates.");
```
They won't actually be sent if this is printed, correct?

denniszollo replied:
affirmative.  Perhaps make it this: "Invalid surveyed position coordinates. No base position msg will be sent."

jacobmcnamee commented on src/track.h:
```diff
@@ -47,7 +47,7 @@ void tracking_channels_missed_update_error(u32 channels_mask);
 /* State management interface */
 bool tracker_channel_available(tracker_channel_id_t id, gnss_signal_t sid);
 bool tracker_channel_init(tracker_channel_id_t id, gnss_signal_t sid,
-                          u32 start_sample_count, float code_phase,
```
remove

ljbade replied:
I noticed this, it didnt match the variable name in track.c's `tracker_channel_init`

jacobmcnamee replied:
Hmm ok. In that case blame Gareth.

swiftnav-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed):
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/887/

ljbade commented:
This PR now depends on https://github.com/swift-nav/piksi_firmware/pull/674

swiftnav-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed):
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/901/

swiftnav-jenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed):
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/918/

ljbade commented:
Merging once travis passes

swiftnav-jenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed):
http://zazu.ci.swiftnav.com/job/piksi_firmware_Pull_Requests/920/
